dpnp/backend/extensions/lapack/evd_batch_common.hpp

```diff
@@ -0,0 +1,174 @@
+    }
+    else if (eig_vecs_shape[2] != eig_vals_shape[0] ||
+             eig_vecs_shape[0] != eig_vals_shape[1])
+    {
+        throw py::value_error(
+            "The shape of 'eig_vals' must be (batch_size, n), "
+            "where batch_size = " +
+            std::to_string(eig_vecs_shape[0]) +
+            " and n = " + std::to_string(eig_vecs_shape[1]));
+    }
+
+    size_t src_nelems(1);
+
+    for (int i = 0; i < eig_vecs_nd; ++i) {
+        src_nelems *= static_cast<size_t>(eig_vecs_shape[i]);
+    }
+
+    if (src_nelems == 0) {
+        // nothing to do
+        return std::make_pair(sycl::event(), sycl::event());
+    }
+
+    dpctl::tensor::validation::CheckWritable::throw_if_not_writable(eig_vecs);
+    dpctl::tensor::validation::CheckWritable::throw_if_not_writable(eig_vals);
+
+    // check compatibility of execution queue and allocation queue
+    if (!dpctl::utils::queues_are_compatible(exec_q, {eig_vecs, eig_vals})) {
+        throw py::value_error(
+            "Execution queue is not compatible with allocation queues");
+    }
+
+    auto const &overlap = dpctl::tensor::overlap::MemoryOverlap();
+    if (overlap(eig_vecs, eig_vals)) {
+        throw py::value_error("Arrays with eigenvectors and eigenvalues are "
+                              "overlapping segments of memory");
+    }
+
+    bool is_eig_vecs_f_contig = eig_vecs.is_f_contiguous();
+    bool is_eig_vals_c_contig = eig_vals.is_c_contiguous();
+    if (!is_eig_vecs_f_contig) {
+        throw py::value_error(
+            "An array with input matrix / output eigenvectors "
+            "must be F-contiguous");
+    }
+    else if (!is_eig_vals_c_contig) {
+        throw py::value_error(
+            "An array with output eigenvalues must be C-contiguous");
+    }
+
+    auto array_types = dpctl_td_ns::usm_ndarray_types();
+    int eig_vecs_type_id =
+        array_types.typenum_to_lookup_id(eig_vecs.get_typenum());
+    int eig_vals_type_id =
+        array_types.typenum_to_lookup_id(eig_vals.get_typenum());
+
+    evd_batch_impl_fn_ptr_t evd_batch_fn =
+        evd_batch_dispatch_table[eig_vecs_type_id][eig_vals_type_id];
+    if (evd_batch_fn == nullptr) {
+        throw py::value_error(
+            "Types of input vectors and result array are mismatched.");
+    }
+
+    char *eig_vecs_data = eig_vecs.get_data();
+    char *eig_vals_data = eig_vals.get_data();
+
+    const std::int64_t batch_size = eig_vecs_shape[2];
+    const std::int64_t n = eig_vecs_shape[1];
+
+    const oneapi::mkl::job jobz_val = static_cast<oneapi::mkl::job>(jobz);
+    const oneapi::mkl::uplo uplo_val =
+        static_cast<oneapi::mkl::uplo>(upper_lower);
+
+    sycl::event evd_batch_ev =
+        evd_batch_fn(exec_q, jobz_val, uplo_val, batch_size, n, eig_vecs_data,
+                     eig_vals_data, depends);
+
+    sycl::event ht_ev = dpctl::utils::keep_args_alive(
+        exec_q, {eig_vecs, eig_vals}, {evd_batch_ev});
+
+    return std::make_pair(ht_ev, evd_batch_ev);
```

Comment:
In reality `evd_batch_ev` is a host task. Would it be safe to return it by the 2nd argument here, because it can be passed later as dependent event to a function from other extension or any function which accepts `depends` keyword? Would it require any additional action, like `py::gil_scoped_release release;` before passing it as dependent event to avoid a dead lock, or it's absolutely safe?